fastcrypto/src/polynomial.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Most of the code below is based on code from https://github.com/celo-org/celo-threshold-bls-rs,
+// modified for our needs.
+//
+
+use crate::error::FastCryptoError;
+use crate::groups::{GroupElement, Scalar};
+use crate::traits::AllowedRng;
+use serde::{Deserialize, Serialize};
+use std::collections::HashSet;
+use std::num::NonZeroU32;
+
+//// Types
+
+/// Indexes of shares (0 is reserved for the secret itself).
+pub type ShareIndex = NonZeroU32;
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct IndexedValue<A> {
+    pub index: ShareIndex,
+    pub value: A,
+}
+
+pub type Eval<A> = IndexedValue<A>;
+
+/// A polynomial that is using a scalar for the variable x and a generic
+/// element for the coefficients.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Poly<C>(Vec<C>);
+
+pub type PrivatePoly<C> = Poly<<C as GroupElement>::ScalarType>;
+pub type PublicPoly<C> = Poly<C>;
+
+//// Vector related operations.
+
+impl<C> Poly<C> {
+    /// Returns the degree of the polynomial
+    pub fn degree(&self) -> usize {
+        // e.g. c_0 + c_1 * x + c_2 * x^2 + c_3 * x^3
+        // ^ 4 coefficients correspond to a 3rd degree poly
+        self.0.len() - 1
+    }
+}
+
+impl<C> From<Vec<C>> for Poly<C> {
+    fn from(c: Vec<C>) -> Self {
+        Self(c)
+    }
+}
+
+impl<C> From<Poly<C>> for Vec<C> {
+    fn from(poly: Poly<C>) -> Self {
+        poly.0
+    }
+}
+
+//// GroupElement operations.
+
+impl<C: GroupElement> Poly<C> {
+    /// Returns a polynomial with the zero element.
+    pub fn zero() -> Self {
+        Self::from(vec![C::zero()])
+    }
+
+    /// Performs polynomial addition in place.
+    pub fn add(&mut self, other: &Self) {
+        // if we have a smaller degree we should pad with zeros
+        if self.0.len() < other.0.len() {
+            self.0.resize(other.0.len(), C::zero())
+        }
+        self.0.iter_mut().zip(&other.0).for_each(|(a, b)| *a += *b)
+    }
+
+    // TODO: Some of the functions/steps below may be executed many times in practice thus cache can be
```

Comment:
Good point, what is a good example in practice for our uses where we'll request at the same index?

Reply:
e.g., you call it when you want to check a partial signature, so assuming some validators are always answering faster than others, you could cache their public keys, either here or on higher level (e.g., the authentication server).